osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs

```diff
@@ -54,5 +60,134 @@ public void TestElapseThenReset()
 
             AddAssert("current time < time at reset", () => gcc.GameplayClock.CurrentTime < timeAtReset);
         }
+
+        [Test]
+        public void TestTrueGameplayRate()
+        {
+            MasterGameplayClockContainer gcc = null;
+
+            AddStep("create container", () =>
+            {
+                var working = CreateWorkingBeatmap(new OsuRuleset().RulesetInfo);
+                working.LoadTrack();
+
+                Add(gcc = new TestMasterGameplayClockContainer(working, 0));
+
+                gcc.Start();
+            });
+
+            AddUntilStep("gameplay clock rate = 1", () => gcc.GameplayClock.Rate == 1f);
+            AddAssert("true gameplay rate = 1", () => gcc.TrueGameplayRate == 1f);
+
+            AddStep("stop gcc", () => gcc.Stop());
+            AddUntilStep("gameplay clock rate = 0", () => gcc.GameplayClock.Rate == 0f);
+            AddAssert("true gameplay rate = 1", () => gcc.TrueGameplayRate == 1f);
+
+            AddStep("change user playback rate to 2", () => gcc.UserPlaybackRate.Value = 2f);
+            AddAssert("gameplay clock rate = 0", () => gcc.GameplayClock.Rate == 0f);
+            AddAssert("true gameplay rate = 1", () => gcc.TrueGameplayRate == 1f);
+
+            AddStep("add 1.5 gameplay freq adjustment", () => gcc.GameplayTrack.AddAdjustment(AdjustableProperty.Frequency, new BindableDouble(1.5f)));
+            AddAssert("gameplay clock rate = 0", () => gcc.GameplayClock.Rate == 0f);
+            AddAssert("true gameplay rate = 1.5", () => gcc.TrueGameplayRate == 1.5f);
+        }
+
+        [Test]
+        public void TestHardwareCorrectionOffsetClock()
+        {
+            OsuConfigManager config = null;
+            WorkingBeatmap working = null;
+            TestMasterGameplayClockContainer gcc = null;
+
+            AddStep("create container", () =>
+            {
+                config = new OsuConfigManager(LocalStorage);
+
+                working = CreateWorkingBeatmap(new OsuRuleset().RulesetInfo);
+                working.LoadTrack();
+
+                Add(new DependencyProvidingContainer
+                {
+                    CachedDependencies = new (Type, object)[] { (typeof(OsuConfigManager), config) },
+                    Child = gcc = new TestMasterGameplayClockContainer(working, 0),
+                });
+
+                gcc.Stop();
+                gcc.Reset();
+            });
+
+            AddStep("set audio offset to 100ms", () => config.SetValue(OsuSetting.AudioOffset, 100.0));
+
+            AddAssert("current time has 100% offset", () => gcc.GameplayClock.CurrentTime == 100);
+
+            AddStep("add 0 track freq adjustment", () => working.Track.AddAdjustment(AdjustableProperty.Frequency, new BindableDouble()));
```

Comment:
I was hoping to use your tests against my PR to see if I can get things working, but this seems wrong? A mod could potentially be adjusting the frequency of the track in a similar way, and *should* be considered as a rate adjust in that case?

Reply:
The behaviours in general have changed in this PR, and the *applicable* adjustments should go to `gcc.GameplayTrack`. This case in this PR is identical to how `pauseFreqAdjust` gets applied.